src/seq_join.rs

```diff
@@ -0,0 +1,361 @@
+{
+    seq_join(active, iterable).try_collect()
+}
+
+#[pin_project]
+pub struct SequentialFutures<I, F>
+where
+    I: Iterator<Item = F> + Send,
+    F: IntoFuture,
+{
+    iter: Fuse<I>,
+    active: VecDeque<Pin<Box<F::IntoFuture>>>,
+}
+
+impl<I, F> Stream for SequentialFutures<I, F>
+where
+    I: Iterator<Item = F> + Send,
+    F: IntoFuture,
+{
+    type Item = F::Output;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+
+        // Draw more values from the input, up to the capacity.
+        while this.active.len() < this.active.capacity() {
+            if let Some(f) = this.iter.next() {
+                this.active.push_back(Box::pin(f.into_future()));
+            } else {
+                break;
+            }
+        }
+
+        if let Some(f) = this.active.front_mut() {
+            if let Poll::Ready(v) = Future::poll(Pin::as_mut(f), cx) {
+                drop(this.active.pop_front());
+                Poll::Ready(Some(v))
+            } else {
+                for f in this.active.iter_mut().skip(1) {
+                    let res = Future::poll(Pin::as_mut(f), cx);
+                    assert!(res.is_pending(), "future resolved out of order");
+                }
+                Poll::Pending
+            }
+        } else {
+            assert!(this.iter.next().is_none());
+            Poll::Ready(None)
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let in_progress = self.active.len();
+        let (lower, upper) = self.iter.size_hint();
+        (
+            lower.saturating_add(in_progress),
+            upper.and_then(|u| u.checked_add(in_progress)),
+        )
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::seq_join::{seq_join, seq_try_join_all};
+    use futures::{
+        future::{lazy, pending, BoxFuture},
+        Future, StreamExt,
+    };
+    use std::{convert::Infallible, iter::once, num::NonZeroUsize};
+
+    async fn immediate(count: u32) {
+        let capacity = NonZeroUsize::new(3).unwrap();
+        let values = seq_join(capacity, (0..count).map(|i| async move { i }))
+            .collect::<Vec<_>>()
+            .await;
+        assert_eq!((0..count).collect::<Vec<_>>(), values);
+    }
+
+    #[tokio::test]
+    async fn within_capacity() {
+        immediate(2).await;
+        immediate(1).await;
+    }
+
+    #[tokio::test]
+    async fn over_capacity() {
+        immediate(10).await;
+    }
+
+    #[tokio::test]
+    #[should_panic(expected = "future resolved out of order")]
+    async fn out_of_order() {
+        let capacity = NonZeroUsize::new(3).unwrap();
+        let unresolved: BoxFuture<'_, u32> = Box::pin(pending());
+        let it = once(unresolved)
+            .chain((0..3_u32).map(|i| -> BoxFuture<'_, u32> { Box::pin(async move { i }) }));
+        drop(seq_join(capacity, it).collect::<Vec<_>>().await);
+    }
+
+    #[cfg(seq_join_stream)]
```

Comment:
This is the same as the comments.  I want streams, but apparently we can't have nice things.